File: cookbook/copilot-sdk/go/recipe/pr-visualization.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"os/exec"
	"regexp"
	"strings"

	"github.com/github/copilot-sdk/go"
)

// ============================================================================
// Git & GitHub Detection
// ============================================================================

func isGitRepo() bool {
	cmd := exec.Command("git", "rev-parse", "--git-dir")
	return cmd.Run() == nil
}

func getGitHubRemote() string {
	cmd := exec.Command("git", "remote", "get-url", "origin")
	output, err := cmd.Output()
	if err != nil {
		return ""
	}

	remoteURL := strings.TrimSpace(string(output))

	// Handle SSH: git@github.com:owner/repo.git
	sshRe := regexp.MustCompile(`git@github\.com:(.+/.+?)(?:\.git)?$`)
	if matches := sshRe.FindStringSubmatch(remoteURL); matches != nil {
		return matches[1]
	}

	// Handle HTTPS: https://github.com/owner/repo.git
	httpsRe := regexp.MustCompile(`https://github\.com/(.+/.+?)(?:\.git)?$`)
	if matches := httpsRe.FindStringSubmatch(remoteURL); matches != nil {
		return matches[1]
	}

	return ""
}

func promptForRepo() string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter GitHub repo (owner/repo): ")
	repo, _ := reader.ReadString('\n')
	return strings.TrimSpace(repo)
}

// ============================================================================
// Main Application
// ============================================================================

func main() {
	repoFlag := flag.String("repo", "", "GitHub repository (owner/repo)")
	flag.Parse()

	fmt.Println("🔍 PR Age Chart Generator\n")

	// Determine the repository
	var repo string

	if *repoFlag != "" {
		repo = *repoFlag
		fmt.Printf("📦 Using specified repo: %s\n", repo)
	} else if isGitRepo() {
		detected := getGitHubRemote()
		if detected != "" {
			repo = detected
			fmt.Printf("📦 Detected GitHub repo: %s\n", repo)
		} else {
			fmt.Println("⚠️  Git repo found but no GitHub remote detected.")
			repo = promptForRepo()
		}
	} else {
		fmt.Println("📁 Not in a git repository.")
		repo = promptForRepo()
	}

	if repo == "" || !strings.Contains(repo, "/") {
		log.Fatal("❌ Invalid repo format. Expected: owner/repo")
	}

	parts := strings.SplitN(repo, "/", 2)
	owner, repoName := parts[0], parts[1]

	// Create Copilot client - no custom tools needed!
	client := copilot.NewClient(copilot.ClientConfig{LogLevel: "error"})

	if err := client.Start(); err != nil {
		log.Fatal(err)
	}
	defer client.Stop()

	cwd, _ := os.Getwd()
	session, err := client.CreateSession(copilot.SessionConfig{
		Model: "gpt-5",
		SystemMessage: copilot.SystemMessage{
			Content: fmt.Sprintf(`
<context>
You are analyzing pull requests for the GitHub repository: %s/%s
The current working directory is: %s
</context>

<instructions>
- Use the GitHub MCP Server tools to fetch PR data
- Use your file and code execution tools to generate charts
- Save any generated images to the current working directory
- Be concise in your responses
</instructions>
`, owner, repoName, cwd),
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	defer session.Destroy()

	// Set up event handling
	session.On(func(event copilot.Event) {
		switch e := event.(type) {
		case copilot.AssistantMessageEvent:
			fmt.Printf("\n🤖 %s\n\n", e.Data.Content)
		case copilot.ToolExecutionStartEvent:
			fmt.Printf("  ⚙️  %s\n", e.Data.ToolName)
		}
	})

	// Initial prompt - let Copilot figure out the details
	fmt.Println("\n📊 Starting analysis...\n")

	prompt := fmt.Sprintf(`
      Fetch the open pull requests for %s/%s from the last week.
      Calculate the age of each PR in days.
      Then generate a bar chart image showing the distribution of PR ages
      (group them into sensible buckets like <1 day, 1-3 days, etc.).
      Save the chart as "pr-age-chart.png" in the current directory.
      Finally, summarize the PR health - average age, oldest PR, and how many might be considered stale.
    `, owner, repoName)

	if err := session.Send(copilot.MessageOptions{Prompt: prompt}); err != nil {
		log.Fatal(err)
	}

	session.WaitForIdle()

	// Interactive loop
	fmt.Println("\n💡 Ask follow-up questions or type \"exit\" to quit.\n")
	fmt.Println("Examples:")
	fmt.Println("  - \"Expand to the last month\"")
	fmt.Println("  - \"Show me the 5 oldest PRs\"")
	fmt.Println("  - \"Generate a pie chart instead\"")
	fmt.Println("  - \"Group by author instead of age\"")
	fmt.Println()

	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print("You: ")
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)

		if input == "" {
			continue
		}
		if strings.ToLower(input) == "exit" || strings.ToLower(input) == "quit" {
			fmt.Println("👋 Goodbye!")
			break
		}

		if err := session.Send(copilot.MessageOptions{Prompt: input}); err != nil {
			log.Printf("Error: %v", err)
		}

		session.WaitForIdle()
	}
}

File: cookbook/copilot-sdk/dotnet/recipe/pr-visualization.cs
#:package GitHub.Copilot.SDK@*
#:property PublishAot=false

using System.Diagnostics;
using GitHub.Copilot.SDK;

// ============================================================================
// Git & GitHub Detection
// ============================================================================

bool IsGitRepo()
{
    try
    {
        var proc = Process.Start(new ProcessStartInfo
        {
            FileName = "git",
            Arguments = "rev-parse --git-dir",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });
        proc?.WaitForExit();
        return proc?.ExitCode == 0;
    }
    catch
    {
        return false;
    }
}

string? GetGitHubRemote()
{
    try
    {
        var proc = Process.Start(new ProcessStartInfo
        {
            FileName = "git",
            Arguments = "remote get-url origin",
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });

        var remoteUrl = proc?.StandardOutput.ReadToEnd().Trim();
        proc?.WaitForExit();

        if (string.IsNullOrEmpty(remoteUrl)) return null;

        // Handle SSH: git@github.com:owner/repo.git
        var sshMatch = System.Text.RegularExpressions.Regex.Match(
            remoteUrl, @"git@github\.com:(.+/.+?)(?:\.git)?$");
        if (sshMatch.Success) return sshMatch.Groups[1].Value;

        // Handle HTTPS: https://github.com/owner/repo.git
        var httpsMatch = System.Text.RegularExpressions.Regex.Match(
            remoteUrl, @"https://github\.com/(.+/.+?)(?:\.git)?$");
        if (httpsMatch.Success) return httpsMatch.Groups[1].Value;

        return null;
    }
    catch
    {
        return null;
    }
}

string? ParseRepoArg(string[] args)
{
    var repoIndex = Array.IndexOf(args, "--repo");
    if (repoIndex != -1 && repoIndex + 1 < args.Length)
    {
        return args[repoIndex + 1];
    }
    return null;
}

string PromptForRepo()
{
    Console.Write("Enter GitHub repo (owner/repo): ");
    return Console.ReadLine()?.Trim() ?? "";
}

// ============================================================================
// Main Application
// ============================================================================

Console.WriteLine("🔍 PR Age Chart Generator\n");

// Determine the repository
var repo = ParseRepoArg(args);

if (!string.IsNullOrEmpty(repo))
{
    Console.WriteLine($"📦 Using specified repo: {repo}");
}
else if (IsGitRepo())
{
    var detected = GetGitHubRemote();
    if (detected != null)
    {
        repo = detected;
        Console.WriteLine($"📦 Detected GitHub repo: {repo}");
    }
    else
    {
        Console.WriteLine("⚠️  Git repo found but no GitHub remote detected.");
        repo = PromptForRepo();
    }
}
else
{
    Console.WriteLine("📁 Not in a git repository.");
    repo = PromptForRepo();
}

if (string.IsNullOrEmpty(repo) || !repo.Contains('/'))
{
    Console.WriteLine("❌ Invalid repo format. Expected: owner/repo");
    return;
}

var parts = repo.Split('/');
var owner = parts[0];
var repoName = parts[1];

// Create Copilot client - no custom tools needed!
await using var client = new CopilotClient(new CopilotClientOptions { LogLevel = "error" });
await client.StartAsync();

var session = await client.CreateSessionAsync(new SessionConfig
{
    Model = "gpt-5",
    SystemMessage = new SystemMessageConfig
    {
        Content = $"""
<context>
You are analyzing pull requests for the GitHub repository: {owner}/{repoName}
The current working directory is: {Environment.CurrentDirectory}
</context>

<instructions>
- Use the GitHub MCP Server tools to fetch PR data
- Use your file and code execution tools to generate charts
- Save any generated images to the current working directory
- Be concise in your responses
</instructions>
"""
    }
});

// Set up event handling
session.On(evt =>
{
    switch (evt)
    {
        case AssistantMessageEvent msg:
            Console.WriteLine($"\n🤖 {msg.Data.Content}\n");
            break;
        case ToolExecutionStartEvent toolStart:
            Console.WriteLine($"  ⚙️  {toolStart.Data.ToolName}");
            break;
    }
});

// Initial prompt - let Copilot figure out the details
Console.WriteLine("\n📊 Starting analysis...\n");

await session.SendAsync(new MessageOptions
{
    Prompt = $"""
      Fetch the open pull requests for {owner}/{repoName} from the last week.
      Calculate the age of each PR in days.
      Then generate a bar chart image showing the distribution of PR ages
      (group them into sensible buckets like <1 day, 1-3 days, etc.).
      Save the chart as "pr-age-chart.png" in the current directory.
      Finally, summarize the PR health - average age, oldest PR, and how many might be considered stale.
    """
});

// Interactive loop
Console.WriteLine("\n💡 Ask follow-up questions or type \"exit\" to quit.\n");
Console.WriteLine("Examples:");
Console.WriteLine("  - \"Expand to the last month\"");
Console.WriteLine("  - \"Show me the 5 oldest PRs\"");
Console.WriteLine("  - \"Generate a pie chart instead\"");
Console.WriteLine("  - \"Group by author instead of age\"");
Console.WriteLine();

while (true)
{
    Console.Write("You: ");
    var input = Console.ReadLine()?.Trim();

    if (string.IsNullOrEmpty(input)) continue;
    if (input.ToLower() is "exit" or "quit")
    {
        Console.WriteLine("👋 Goodbye!");
        break;
    }

    await session.SendAsync(new MessageOptions { Prompt = input });
}
